test/lit/passes/type-ssa.wast

```diff
@@ -0,0 +1,221 @@
+    (drop
+      (struct.new $struct
+        (i32.const 100)
+      )
+    )
+  )
+)
+
+;; Some of these are uninteresting and should not get a new type.
+(module
+  ;; CHECK:      (type $anyref_dataref_=>_none (func (param anyref dataref)))
+
+  ;; CHECK:      (type $struct (struct (field anyref)))
+  ;; NOMINAL:      (type $anyref_dataref_=>_none (func (param anyref dataref)))
+
+  ;; NOMINAL:      (type $struct (struct (field anyref)))
+  (type $struct (struct_subtype (field (ref null any)) data))
+
+  ;; CHECK:      (rec
+  ;; CHECK-NEXT:  (type $struct$1 (struct_subtype (field anyref) $struct))
+
+  ;; CHECK:       (type $struct$2 (struct_subtype (field anyref) $struct))
+
+  ;; CHECK:       (type $struct$3 (struct_subtype (field anyref) $struct))
+
+  ;; CHECK:      (func $foo (type $anyref_dataref_=>_none) (param $any anyref) (param $data dataref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new_default $struct$1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new $struct$2
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new $struct
+  ;; CHECK-NEXT:    (local.get $any)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new $struct$3
+  ;; CHECK-NEXT:    (local.get $data)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block ;; (replaces something unreachable we can't emit)
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (unreachable)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (unreachable)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  ;; NOMINAL:      (type $struct$1 (struct_subtype (field anyref) $struct))
+
+  ;; NOMINAL:      (type $struct$2 (struct_subtype (field anyref) $struct))
+
+  ;; NOMINAL:      (type $struct$3 (struct_subtype (field anyref) $struct))
+
+  ;; NOMINAL:      (func $foo (type $anyref_dataref_=>_none) (param $any anyref) (param $data dataref)
+  ;; NOMINAL-NEXT:  (drop
+  ;; NOMINAL-NEXT:   (struct.new_default $struct$1)
+  ;; NOMINAL-NEXT:  )
+  ;; NOMINAL-NEXT:  (drop
+  ;; NOMINAL-NEXT:   (struct.new $struct$2
+  ;; NOMINAL-NEXT:    (ref.null none)
+  ;; NOMINAL-NEXT:   )
+  ;; NOMINAL-NEXT:  )
+  ;; NOMINAL-NEXT:  (drop
+  ;; NOMINAL-NEXT:   (struct.new $struct
+  ;; NOMINAL-NEXT:    (local.get $any)
+  ;; NOMINAL-NEXT:   )
+  ;; NOMINAL-NEXT:  )
+  ;; NOMINAL-NEXT:  (drop
+  ;; NOMINAL-NEXT:   (struct.new $struct$3
+  ;; NOMINAL-NEXT:    (local.get $data)
+  ;; NOMINAL-NEXT:   )
+  ;; NOMINAL-NEXT:  )
+  ;; NOMINAL-NEXT:  (drop
+  ;; NOMINAL-NEXT:   (block ;; (replaces something unreachable we can't emit)
+  ;; NOMINAL-NEXT:    (drop
+  ;; NOMINAL-NEXT:     (unreachable)
+  ;; NOMINAL-NEXT:    )
+  ;; NOMINAL-NEXT:    (unreachable)
+  ;; NOMINAL-NEXT:   )
+  ;; NOMINAL-NEXT:  )
+  ;; NOMINAL-NEXT: )
+  (func $foo (param $any (ref null any)) (param $data (ref null data))
```

Comment:
It would be great if you could use `array` or some other type instead of `data`, since I'm planning to remove `data` soon.

Reply:
Done.